spfx doctor: add gulp-cli checks

DonKirkham opened the pull request:
### Add gulp-cli checks to spfx doctor. Closes #3541

Closes #3541

Added method to check for gulp-cli

Updated method that checks gulp to just check if it is present, which triggers failure and fix (updated below)

Updated the prerequisites section as follows:
- Added section for gulp-cli to each SPFx version
--  Any version of gulp-cli is valid since all versions of SPFx use gulp ^3 || ^4 
-- Fix for missing gulp-cli is to install latest "npm i -g gulp-cli"
- Updated the gulp section for each version of SPFx
-- I left the version information the same, even though we don't compare versions
-- changed the fix to uninstall gulp "npm un -g gulp"

I'm not sure where lines 3, 4 & 6 came from. I did not intentionally add them. They are not needed for the changes I made.

waldekmastykarz commented on src/m365/spfx/commands/spfx-doctor.ts:
```diff
@@ -82,7 +86,11 @@ class SpfxDoctorCommand extends AnonymousCommand {
       gulp: {
         range: '^3',
-        fix: 'npm i -g gulp@3'
+        fix: 'npm un -g gulp'
```
We shouldn't be proposing uninstalling gulp. That's beyond the scope of this issue.

DonKirkham replied:
My understanding is that the whole reason for the switch to gulp-cli was that having gulp installed globally can lead to incompatibility issues, so the recommendation is to NOT install gulp. Moving forward, hopefully this will not be an issue, but there is a lot of documentation out on the interwebs that still has devs installing gulp instead of gulp-cli.

DonKirkham replied:
Maybe we can change the message (line 727) to say "gulp should be removed in favor of gulp-cli" instead. I tried to look at checking for both, then customizing the messaging, but it was more of a change than I was willing to make on my first attempt. Keeping the checks separate was simpler.

waldekmastykarz replied:
Correct, our recommendation is to suggest installing gulp-cli, but I'm not sure if we should go as far as suggesting uninstalling global gulp. If that's the consensus, then we should implement it in such a way that it applies to all versions at once, because it's the same check no matter which version you have installed.

DonKirkham replied:
Not sure what you mean by "all versions at once." All versions of gulp-cli are compatible with gulp 3  & 4, and all versions of SPFx require gulp 3 or 4, so the changes I made to the command should be accurate, regardless of which version of SPFx you install.

waldekmastykarz replied:
For installing `gulp-cli` we want to install a specific version, that's compatible with the version of SPFx that you want to use. For uninstalling `gulp`, we don't care which version you have installed, as we'll recommend uninstalling it anyway.

waldekmastykarz commented on src/m365/spfx/commands/spfx-doctor.ts:
```diff
@@ -82,7 +86,11 @@ class SpfxDoctorCommand extends AnonymousCommand {
+      },
+      gulpCli: {
+        range: '^1 || ^2',
+        fix: 'npm i -g gulp-cli'
```
Since we require v1 or v2, we should update the install command, to `npm i -g gulp-cli@2` so that it installs v2 rather than the latest currently available version.

DonKirkham replied:
Got it. So when gulp-cli@3 comes out (if ever), we would need to update because doctor would report an error.  I'll fix it

waldekmastykarz replied:
My understanding is that a specific version of SPFx requires a specific version(s) of gulp-cli, so we should be explicit about them, because in the future new versions might be available that aren't compatible with older versions of SPFx.

DonKirkham replied:
Fixed in latest commit

waldekmastykarz commented on src/m365/spfx/commands/spfx-doctor.ts:
```diff
@@ -534,6 +635,7 @@ class SpfxDoctorCommand extends AnonymousCommand {
       .then(_ => this.checkSharePointCompatibility(spfxVersion, prerequisites, args, fixes, logger))
       .then(_ => this.checkNodeVersion(prerequisites, fixes, logger))
       .then(_ => this.checkYo(prerequisites, fixes, logger))
+      .then(_ => this.checkGulpCli(prerequisites, fixes, logger))
```
`checkGulpCli` should replace `checkGulp`. There's no need for us to control both

DonKirkham replied:
See comment on uninstalling gulp. I think doctor should recommend gulp be uninstalled if it is present. I changed the code to ignore the version of gulp, just recommend removing it. Wasn't sure how to do this other than the way I did it with the checkGulp function

DonKirkham replied:
I tried to remove the gulp.range lines since we do not use them, but that causes an error from the interface. If we choose to leave this check in, should I just make the versions an empty string?

waldekmastykarz replied:
Which interface do you mean exactly?

DonKirkham replied:
The VersionCheck interface had range listed as required.

```
interface VersionCheck {
  /**
   * Required version range in semver
   */
  range: string;
  /**
   * What to do to fix it if the required range isn't met
   */
  fix: string;
}
```

Should I just change all of the "range" properties to empty strings in the gulp objects? For example,
```
private readonly versions: { [version: string]: SpfxVersionPrerequisites } = {
'1.0.0': {
	        gulp: {
-->	          range: '',
	          fix: 'npm un -g gulp'
	        },
			...
```

waldekmastykarz replied:
Got it. Like I said, there's no need to list gulp in each every version separately, because the check and the rule to fix it is the same for all versions. So we can remove it from the versions object altogether and simply run as one of the checks on line 638, right before checking gulp-cli

DonKirkham replied:
Update made, removing gulp from the versions object

waldekmastykarz commented on src/m365/spfx/commands/spfx-doctor.ts:
```diff
@@ -1,6 +1,9 @@
 import * as chalk from 'chalk';
 import * as child_process from 'child_process';
+import { resolve } from 'dns';
```
Why do we need these imports on line 3, 4 and 6?

DonKirkham replied:
I am not sure why those got added. I did not do it directly, but suspect it happened as I was trying to sync my fork to pnp. I think those were taken out at some point?  I will remove them

waldekmastykarz replied:
Let's indeed ensure that your PR doesn't contain anything that we shouldn't include.

waldekmastykarz reviewed (changes requested):
Nicely done. Let's do a few adjustments before we merge it.

thechriskent commented:
Gulp's [getting started documentation](https://github.com/gulpjs/gulp/blob/master/docs/getting-started/1-quick-start.md) says "If you've previously installed gulp globally, run npm rm --global gulp before following these instructions."

So, if we're recommending the installation of gulp-cli then we should pass along gulp's recommendation to remove gulp first.

waldekmastykarz commented:
It seems that we've got some broken tests. Would you mind having a look before we process the PR @DonKirkham?

DonKirkham commented:
I ran "npm test" and it reported an error with tabs vs spaces. I fixed that. It is also showing a code coverage error, but it looks like that is not in the files I worked with.  Not sure if I fixed everything needed, but I gave it a shot.

DonKirkham commented:
I updated the tests, changing references for gulp to gulp-cli. I see three issues:

1. Lint reports that there are 3 import statements in spfx-doctor.ts that are not used. I have tried to delete them, but when I save, they come back. I think it is a merge issue, but can't find a fix.
2. The cov test reports less than 100% coverage in spfx-doctor.ts. The reported outage is that there is no test to cover what happens when gulp is found, lines 630-632. I am not sure how to write this new test, so I need some guidance.
3. The cov test is reporting the following error, which is in an area I have not touched.

```
 1) utils/md
       converts code fences:

      AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  assert(cliCompletionClinkUpdateHelpPlain.includes('  cli completion clink update > m365.lua'))

      + expected - actual

      -false
      +true
```
 Thanks for the help getting this finished.

waldekmastykarz commented on src/m365/spfx/commands/spfx-doctor.ts:
```diff
@@ -603,16 +605,27 @@ class SpfxDoctorCommand extends AnonymousCommand {
+      });
+  }
+
   private checkGulp(prerequisites: SpfxVersionPrerequisites, fixes: string[], logger: Logger): Promise<void> {
```
Because we don't use prerequisites here, let's remove them from the method's signature to avoid unnecessary arguments

waldekmastykarz commented on src/m365/spfx/commands/spfx-doctor.ts:
```diff
@@ -731,14 +744,14 @@ class SpfxDoctorCommand extends AnonymousCommand {
           } = JSON.parse(responseString);
           if (packageInfo.dependencies &&
-            packageInfo.dependencies[packageName]) {
+                          packageInfo.dependencies[packageName]) {
```
Let's remove this unnecessary indenting

waldekmastykarz commented on src/m365/spfx/commands/spfx-doctor.ts:
```diff
@@ -770,12 +783,12 @@ class SpfxDoctorCommand extends AnonymousCommand {
     const primarySupported: boolean = process.platform !== 'win32' ||
-      process.env.CI === 'true' ||
-      process.env.TERM === 'xterm-256color';
+                          process.env.CI === 'true' ||
```
Let's remove this unnecessary indenting

waldekmastykarz reviewed (approved):
Very nicely done! I've done a few cosmetic changes when merging the PR 👏

waldekmastykarz commented:
Merged manually. Thank you and welcome to the team! 👏 🎉

DonKirkham commented:
> Merged manually. Thank you and welcome to the team! 👏 🎉

Thanks. Sorry about the weird spacing issues. Something in my VS Code settings causes this, but I thought I had fixed them all. Since this is merged, can I delete my branch?

waldekmastykarz commented:
No problem, and yes, you're all good to delete your branch.
